junipervpn/vpn.py

```diff
@@ -65,6 +68,110 @@ def hotp(key):
     counter = int2beint64(int(time.time()) / 30)
     return dec(hmac.new(key, counter, hashlib.sha256).digest(), 6)
 
+
+class NetworkMonitorThread(threading.Thread):
+    """
+    :param rate: Number of packets emitted per second
+    :type rate: float
+
+    :param on_disconnect: Callable to call when a network disconnection is
+        detected.
+    :type on_disconnect: collections.abc.Callable
+
+    :param host: Host to ping.
+    :type host: str
+
+    :param ping: Ping command to use.
+    :type ping: str
+
+    :param ping_rate: Ping rate in packet/s
+    :type ping_rate: float
+
+    :param ping_duration: Duration of each ping command. A short-enough interval is
+        preferred, so that the script can react quickly to network outage
+    :type ping_duration: int
+
+    :param connect_timeout: Time to wait in seconds between reconnection
+        attempts.
+    :type connect_timeout: float
+
+    :param daemon: Passed to threading.Thread.
+    :type daemon: bool
+    """
+    def __init__(self,
+        on_disconnect,
+        host='google.com',
+        ping='ping',
+        connect_timeout=10,
+        ping_rate=1,
+        ping_duration=2,
+        daemon=True,
+    ):
+        self.host = host
+        self.ping = ping
+        self.ping_rate = ping_rate
+        self.ping_duration = ping_duration
+        self.connect_timeout = connect_timeout
+        self.on_disconnect = on_disconnect
+
+        self._stop = threading.Event()
+        super().__init__(name='ping_network_monitor', daemon=daemon)
+
+    def run(self):
+        interval = 1 / self.ping_rate
+        # interval < 0.2 is not allowed as regular user, so best avoided
+        interval = interval if interval >= 0.2 else 0.2
+
+        deadline = int(math.ceil(self.ping_duration))
+        cmd = ['ping', '-q', '-w', str(deadline), '-i', str(interval), '--', self.host]
```

Comment:
Hmmm... I was kind of expecting this to be pinging through the VPN, but this is going over the normal network.

Is this intended to only work when openconnect has created a tun device so all traffic is going through the VPN?

Or I guess if you're using ocproxy or similar, this will reconnect after network outages, but won't reconnect if the VPN crashes "independently". (e.g. if it dies from a network disconnect shorter than the ping interval).

Reply:
Ideally it would work in all situations. It seems that openconnect is good at reconnecting from a VPN crash but not necessarily after a proper network outage. This means that the ping packets should be sent through the VPN to get the best results.

Since ICMP cannot go through a socks proxy it seems, maybe we should use an HTTP request instead of ping, or at least let users change the ping command to anything else they want (CLI/config file), using something like that:
https://stackoverflow.com/questions/5274934/use-ping-through-socks-server

Reply:
Yeah - it looks like `httping` would work?

It's a bit crude but we could scan the `--script` argument for ocproxy's `-D` argument (that's the SOCKS5 proxy), then pass that as `httping`'s proxy argument? That would then work for both cases.

The `--script-tun` argument should tell us whether all traffic is going through the VPN (in which case `ping` is fine), or whether we need to use `httping`. We can then enable the thread intelligently based on whether we can detect the HTTP proxy and whether `httping` is installed or not.